services/supervisor/src/transactions.rs

```diff
@@ -413,6 +447,150 @@ impl SupervisorInterface<ExecutionContext<'_>> for Supervisor {
+            // returning an error, we will just mark this migration as failed.
+            drop(schema);
+            let supervisor_extensions = context.supervisor_extensions();
+            let result = supervisor_extensions
+                .initiate_migration(request.new_artifact.clone(), request.service.as_ref());
+
+            // Check whether migration started successfully.
+            if let Err(error) = result {
+                // Migration failed even before start, softly mark it as failed.
+                let initiate_rollback = false;
+                return self.fail_migration(context, request, error, initiate_rollback);
+            }
+
+            // Migration started. Check if migration is fast-forward.
+            let instance = get_instance_by_name(&context, request.service.as_ref())
+                .expect("BUG: Instance disappeared");
+            if state.version != *instance.data_version() {
+                // Migration is fast-forward, complete it immediately.
+                // No agreement needed, since nodes which will behave differently will obtain
+                // different blockchain state hash and will be excluded from consensus.
+                log::trace!("Applied fast-forward migration with request {:?}", request);
+
+                let mut schema = SchemaImpl::new(context.service_data());
+
+                // Update the state of a migration.
+                let mut state = schema
+                    .migration_states
+                    .get(&request)
+                    .expect("BUG: Migration succeed, but does not have a stored state");
+                state.update(AsyncEventState::Succeed, instance.data_version().clone());
+                schema.migration_states.put(&request, state);
+
+                // Remove the migration from the list of pending.
+                schema.pending_migrations.remove(&request);
+            }
+        }
+        Ok(())
+    }
+
+    fn report_migration_result(
+        &self,
+        context: ExecutionContext<'_>,
+        result: MigrationResult,
+    ) -> Self::Output {
+        // Verifies that transaction author is validator.
+        let author = get_validator(&context)?;
+
+        result
+            .request
+            .new_artifact
+            .validate()
+            .map_err(|e| ArtifactError::InvalidArtifactId.with_description(e))?;
+
+        let core_schema = context.data().for_core();
+        let current_height = core_schema.height();
+
+        let schema = SchemaImpl::new(context.service_data());
+
+        // Verify that this migration is registered.
+        schema
+            .migration_states
+            .get(&result.request)
+            .ok_or(MigrationError::MigrationRequestNotRegistered)?;
+
+        // Check if migration already failed.
+        if schema
+            .migration_states
+            .get(&result.request)
+            .map(|state| state.is_failed())
+            .unwrap_or_default()
```

Comment:
Nit: IMO, `unwrap_or(false)` would be more readable. Also, isn't this whole thing redundant? You can bind a migration state from the previous check to a local variable (say, `state`) and simply check `if state.is_failed()` here.